src/VisualStudio/Core/Def/Implementation/Diagnostics/DiagnosticProgressReporter.cs

```diff
@@ -1,28 +1,32 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
 using System.ComponentModel.Composition;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.SolutionCrawler;
 using Microsoft.VisualStudio.Shell;
 using Microsoft.VisualStudio.TaskStatusCenter;
+using Roslyn.Utilities;
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.Diagnostics
 {
     [Export(typeof(DiagnosticProgressReporter))]
-    internal class DiagnosticProgressReporter
+    internal sealed class DiagnosticProgressReporter
     {
+        private static readonly TimeSpan s_minimumInterval = TimeSpan.FromMilliseconds(15);
+
         private readonly IVsTaskStatusCenterService _taskCenterService;
         private readonly IDiagnosticService _diagnosticService;
         private readonly TaskHandlerOptions _options;
 
-        // this fields are never accessed concurrently
-        private TaskCompletionSource<bool> _currentTask;
-        private string _lastMessage;
+        // these fields are never accessed concurrently
+        private TaskCompletionSource<VoidResult> _currentTask;
+        private object _lastReportedDocumentOrProject;
+        private DateTime _lastTimeReported;
```

Comment:
:bulb: Use `DateTimeOffset` instead of `DateTime`. This is similar to the general rule to try and use `TimeSpan` instead of `int`. Even though they are generally interchangeable in the underlying APIs, one option is much harder to misuse than the other.